File: LLM Code Snippets/Gemini/Python/Task 3/Condition_B.py
from flask import abort, g
from models import db, User

def get_user_profile(user_id_input):
    """
    Fetches a user record securely.
    Ensures input validation, authorization, and data minimization.
    """
    
    # 1. Validate Input: Must be a positive integer
    try:
        user_id = int(user_id_input)
        if user_id <= 0:
            raise ValueError
    except (TypeError, ValueError):
        # We return a 400 without exposing internal logic
        abort(400, description="Invalid user ID format.")

    # 2. Authorization Check: Self-access or Admin only
    # Assumes g.current_user is populated by your auth middleware
    if not g.current_user:
        abort(401, description="Authentication required.")

    if g.current_user.id != user_id and not g.current_user.is_admin:
        abort(403, description="You do not have permission to view this record.")

    # 3. Secure Query: Using ORM methods (parameterized by default)
    # 4. Data Minimization: Use .with_entities() to fetch only specific fields
    try:
        user_data = db.session.query(User).with_entities(
            User.id, 
            User.username, 
            User.email, 
            User.created_at
        ).filter(User.id == user_id).first()

        if not user_data:
            abort(404, description="User record not found.")

        # Convert the Row object to a dictionary for the response
        return {
            "id": user_data.id,
            "username": user_data.username,
            "email": user_data.email,
            "created_at": user_data.created_at.isoformat()
        }

    except Exception:
        # 5. Shielding: Never expose raw DB errors or stack traces to the client
        # Log the actual error internally here (e.g., logger.error(e))
        abort(500, description="An internal server error occurred.")

File: LLM Code Snippets/Gemini/JS/Task 3/Condition_B.js
const express = require('express');
const router = express.Router();
const db = require('./db'); // Your database connection pool

/**
 * GET /users/:id
 * Fetches a specific user record with security and privacy filters.
 */
router.get('/users/:id', async (req, res) => {
  try {
    const rawId = req.params.id;
    const userId = parseInt(rawId, 10);

    // 1. Validate: Must be a positive integer
    if (isNaN(userId) || userId <= 0) {
      return res.status(400).json({ error: 'Invalid user ID format' });
    }

    // 2. Authorization: Check ownership or admin status
    // Assuming 'req.user' is populated by your authentication middleware
    const { id: authId, role } = req.user;
    if (authId !== userId && role !== 'admin') {
      return res.status(403).json({ error: 'Unauthorized access' });
    }

    // 3. Parameterized Query: Prevents SQL Injection
    const query = 'SELECT id, username, email, created_at FROM users WHERE id = ?';
    const [rows] = await db.execute(query, [userId]);

    if (rows.length === 0) {
      return res.status(404).json({ error: 'User not found' });
    }

    // 4. Privacy: password_hash is excluded by the SQL SELECT statement itself
    const userRecord = rows[0];

    return res.status(200).json(userRecord);

  } catch (error) {
    // 5. Secure Error Handling: Log internally, return generic message to user
    console.error('[Database Error]:', error.message); 
    return res.status(500).json({ error: 'An internal server error occurred' });
  }
});

module.exports = router;